### PR TITLE
Fancy throttle

### PR DESCRIPTION
**Pull request by 3noch**

### CODE REVIEW

**Comment by lspitzner** [comment]
This reminds me of https://github.com/reflex-frp/reflex/issues/184 which has some similar constructs.

Also, at a glance I wonder if there is much use to using Semigroup if you are not forcing the intermediate value.

**Comment by ElvishJerricco** [comment]
That *is* a fancy throttle! Nice!

**Comment by lspitzner** [comment]
Does this behave as expected for a delay of 5s and a example input signal that fires once every second? If I am reading things correctly, the delayed signals are not filtered, meaning that after 5 seconds, the delayed signal fires every second too, which turns immediate mode back on at the same rate. So you get 1 output, 5s pause, 4 merged outputs, and then every (other?) second some output. Whether it is every second or every other probably is not deterministic due to continuous-time/fuzzy delays.

I'd expect proper 5s intervals of silence, as you'd get if only the non-immediate-mode-entering signals trigger, after a delay, the back-to-immediate-mode switch.

**Comment by ali-abrar** [comment]
@3noch (maybe a question for James) do we have a test that would answer @lspitzner's question?

We should definitely figure out how this interacts/overlaps with #184 before merging. 

@cgibbard may also have some thoughts as the throttle code here seems to be an upgraded version of something he wrote (that somehow never made it to develop..) https://github.com/reflex-frp/reflex/blob/4bdf22876975be3cf53deb6471d746f843569d4d/src/Reflex/Time.hs#L243

**Comment by 3noch** [comment]
cc @xplat

**Comment by xplat** [comment]
> Does this behave as expected for a delay of 5s and a example input signal that fires once every second? If I am reading things correctly, the delayed signals are not filtered, meaning that after 5 seconds, the delayed signal fires every second too, which turns immediate mode back on at the same rate. So you get 1 output, 5s pause, 4 merged outputs, and then every (other?) second some output. Whether it is every second or every other probably is not deterministic due to continuous-time/fuzzy delays.
> 
> I'd expect proper 5s intervals of silence, as you'd get if only the non-immediate-mode-entering signals trigger, after a delay, the back-to-immediate-mode switch.

I had exactly the same question after reading the code for `throttle`.  @cgibbard gave me the answer -- the delay is fired based on the *output* events, not the *input* events.  So the first output, 5s pause, 4 merged outputs, and then nothing until 5 seconds after the 4 merged outputs.

It's all twisty, but it works.

**Comment by lspitzner** [comment]
Right, i should have noticed that. sorry. Looks fine to me!

(and regarding functionality overlap: I think it is worth having this function in addition to the #184 ones. Even if you could express fancy-throttle in terms of `gateGather`, as I suspect you can.)

**Comment by sboosali** [comment]
should we add that clarification to the documentation?

**Comment by ali-abrar** [comment]
Given that the question has come up a couple times, I think we should

**Comment by luigy** [comment]
pinging @cgibbard since I believe you were somewhat involved in this?

File: src/Reflex/Time.hs [hunk 1]
```diff
@@ -35,6 +35,7 @@ import Data.Time.Clock
 import Data.Typeable
 import GHC.Generics (Generic)
 import System.Random
+import Data.Semigroup
```
Comment (3noch) [inline src/Reflex/Time.hs]: The imports were alphabetized prior to this change

**Comment by tomsmalley** [comment]
`throttleBatchWithLag pure` causes a hang which leads to a stack overflow. Granted, the docs say that the "lag" function should construct an event, but we should at least give a stronger warning there.

**Pull request merged**